csi/src/node/nvmfutil.rs

```diff
@@ -0,0 +1,179 @@
+    // to uuid string value.
+    // will return a garbage string if passed in garbage,
+    // but that is okay for now since this is only used
+    // to find devices with a matching uuid and nothing
+    // will match the garbage.
+    let mut frags = s.split('-').collect::<Vec<_>>();
+    while frags.len() > 5 {
+        frags.remove(0);
+    }
+    itertools::join(&frags, "-")
+    // } hack
+}
+
+fn wait_for_path_to_exist(uuid: String, max_retries: i32) -> Option<String> {
+    let second = time::Duration::from_millis(1000);
+    let mut retries: i32 = 0;
+    let now = time::Instant::now();
+    while retries < max_retries {
+        thread::sleep(second);
+
+        if let Ok(path) = find_nvmf_device_by_uuid(&uuid) {
+            trace!(
+                "wait_for_path_to_exist {} {} success for elapsed time is {:?}",
+                uuid,
+                path,
+                now.elapsed()
+            );
+            return Some(path);
+        }
+        retries += 1;
+    }
+    debug!("wait_for_path_to_exist timed out after {:?}", now.elapsed());
+    None
+}
+
+fn nvmeadm_attach_disk(
+    ip_addr: &str,
+    port: u32,
+    nqn: &str,
+    uri: &str,
+) -> Result<String, String> {
+    trace!(
+        "nvmeadm_attach_disk ip_addr={} port={} nqn={}",
+        ip_addr,
+        port,
+        nqn
+    );
+
+    let result = nvmeadm::nvmf_discovery::connect(ip_addr, port, nqn);
+    if let Err(e) = result {
+        let mut in_progress = false;
+        if let Some(ioerr) = e.downcast_ref::<io::Error>() {
+            if let Some(errcode) = ioerr.raw_os_error() {
+                if errcode == 114 {
+                    in_progress = true;
+                    debug!("nvmeadm operation already in progress for {}", uri);
+                }
+            }
+        }
+        if !in_progress {
+            debug!("nvmeadm connect failed error {} {:?}", uri, e);
+            return Err(format!("{}", e));
+        }
+    }
+
+    match wait_for_path_to_exist(uuid_from_str(uri), 10) {
+        Some(path) => Ok(path),
+        _ => {
+            debug!("nvmeadm nvmf device path not found.");
+            Err("No path for nvme device".to_string())
+        }
+    }
+}
+
+pub fn nvmeadm_detach_disk(nqn: &str) -> Result<(), String> {
+    match nvmeadm::nvmf_discovery::disconnect(&nqn) {
+        Ok(_) => {
+            trace!("nvmf disconnected {}", nqn);
+            Ok(())
+        }
+        Err(e) => {
+            debug!("nvmf disconnect {} FAILED.", nqn);
+            Err(format!("{}", e))
+        }
+    }
+}
+
+pub fn nvmf_attach_disk(nvmf_uri: &str) -> Result<String, String> {
+    trace!("nvmf_attach_disk {}", nvmf_uri);
+
+    if let Some(path) = wait_for_path_to_exist(uuid_from_str(nvmf_uri), 1) {
+        return Ok(path);
+    }
+
+    if let Ok(url) = url::Url::parse(nvmf_uri) {
+        if url.scheme() == "nvmf" {
+            let tokens =
+                url.path_segments().map(|c| c.collect::<Vec<_>>()).unwrap();
```

Comment:
the map() is not really necessary here: the following IMO is cleaner and clearer:
let tokens: Vec<&str> = url.path_segments().unwrap().collect()

Reply:
https://github.com/openebs/Mayastor/pull/302/commits/103a309bb8f0d25e4b4cf62544a0d24d400cd998

Reply:
https://github.com/openebs/Mayastor/pull/302/commits/24fc1bb5d399e95874126ee0fdc1bae0b9c1d1a3